numba/np/arrayobj.py

```diff
@@ -1690,67 +1699,74 @@ def fancy_setslice(context, builder, sig, args, index_types, indices):
         with builder.if_then(shape_error, likely=False):
             msg = "cannot assign slice from input of different size"
             context.call_conv.return_user_exc(builder, ValueError, (msg,))
-
-        def src_getitem(source_indices):
-            idx, = source_indices
-            getitem_impl = context.get_function(
-                operator.getitem,
-                signature(src_dtype, srcty, types.intp),
-            )
-            return getitem_impl(builder, (src, idx))
-
-        def src_cleanup():
-            pass
-
+        getitem_impl = context.get_function(
+            operator.getitem,
+            signature(src_dtype, srcty, types.intp))
     else:
         # Source is a scalar (broadcast or not, depending on destination
         # shape).
         src_dtype = srcty
+        src_type = 'scalar'
 
-        def src_getitem(source_indices):
-            return src
-
-        def src_cleanup():
-            pass
-
-    zero = context.get_constant(types.uintp, 0)
-    # Loop on destination and copy from source to destination
-    dest_indices, counts = indexer.begin_loops()
-
-    # Source is iterated in natural order
-
-    # Counts represent a counter for the number of times a specified axis
-    # is being accessed, during setitem they are used as source
-    # indices
-    counts = list(counts)
+    if src_type == 'buffer':
```

Comment:
I don't think that this path in its present form can be supported on CUDA as the `flat_imp*` function compiled below relies on:
* NumPy functions
* CPU only implementations for `reshape`
* returns an array
* needs the NRT

I'm not sure of the best way to "detect" this, immediate thoughts are to either require the NRT and raise if the context doesn't have it, or alternatively declare these functions as overloads targetting CPU only which will then fail for CUDA users. CC @gmarkall do you have an opinion on this?